File: cli/src/commands/review.rs
use clap::Subcommand;

use crate::client::CowikiClient;
use crate::error::CliError;
use crate::output;
use colored::Colorize;

#[derive(Subcommand)]
pub enum ReviewCommand {
    /// List review submissions
    List {
        /// Filter by status (pending/approved/rejected)
        #[arg(long)]
        status: Option<String>,
    },
    /// Show review details with diffs
    Show {
        /// Submission ID
        id: String,
    },
    /// Approve a submission
    Approve {
        /// Submission ID
        id: String,
    },
    /// Reject a submission
    Reject {
        /// Submission ID
        id: String,
    },
}

pub async fn run(
    client: &CowikiClient,
    cmd: ReviewCommand,
    json: bool,
) -> Result<(), CliError> {
    match cmd {
        ReviewCommand::List { status } => list_reviews(client, status, json).await,
        ReviewCommand::Show { id } => show_review(client, &id, json).await,
        ReviewCommand::Approve { id } => approve_review(client, &id, json).await,
        ReviewCommand::Reject { id } => reject_review(client, &id, json).await,
    }
}

async fn list_reviews(
    client: &CowikiClient,
    status: Option<String>,
    json: bool,
) -> Result<(), CliError> {
    let mut submissions = client.list_reviews().await?;

    if let Some(ref s) = status {
        submissions.retain(|sub| sub.status.eq_ignore_ascii_case(s));
    }

    if submissions.is_empty() {
        output::print_info("no reviews found");
        return Ok(());
    }

    if json {
        println!(
            "{}",
            serde_json::to_string_pretty(&submissions).map_err(|e| {
                CliError::Unexpected(format!("failed to serialize reviews: {e}"))
            })?
        );
        return Ok(());
    }

    let mut table = comfy_table::Table::new();
    table.set_header(vec!["ID", "USER", "STATUS", "SUMMARY", "CREATED"]);
    for s in &submissions {
        let status_colored = match s.status.as_str() {
            "pending" => s.status.yellow().to_string(),
            "approved" => s.status.green().to_string(),
            "rejected" => s.status.red().to_string(),
            _ => s.status.clone(),
        };
        table.add_row(vec![
            &s.id,
            &s.user_id,
            &status_colored,
            &s.summary,
            &s.created_at,
        ]);
    }
    output::print_table(table);

    Ok(())
}

async fn show_review(client: &CowikiClient, id: &str, json: bool) -> Result<(), CliError> {
    let detail = client.get_review(id).await.map_err(|e| match &e {
        CliError::Api { status: 404, .. } => {
            CliError::Unexpected(format!("review not found: \"{id}\""))
        }
        _ => e,
    })?;

    if json {
        println!(
            "{}",
            serde_json::to_string_pretty(&detail).map_err(|e| {
                CliError::Unexpected(format!("failed to serialize review: {e}"))
            })?
        );
        return Ok(());
    }

    let sub = &detail.submission;
    let status_colored = match sub.status.as_str() {
        "pending" => sub.status.yellow(),
        "approved" => sub.status.green(),
        "rejected" => sub.status.red(),
        _ => sub.status.normal(),
    };

    println!("{} {}", "Submission:".bold(), sub.id);
    println!("  User:     {}", sub.user_id);
    println!("  Status:   {status_colored}");
    println!("  Summary:  {}", sub.summary);
    println!("  Branch:   {}", sub.source_branch);
    println!("  Created:  {}", sub.created_at);
    if let Some(ref reviewer) = sub.reviewed_by {
        println!("  Reviewed by: {reviewer}");
    }
    if let Some(ref reviewed_at) = sub.reviewed_at {
        println!("  Reviewed at: {reviewed_at}");
    }
    println!();

    if detail.diffs.is_empty() {
        output::print_info("no diffs");
    } else {
        for diff in &detail.diffs {
            println!("{} {}", "---".yellow(), diff.path);
            if let Some(ref old) = diff.old_content {
                for line in old.lines() {
                    println!("{}{line}", "- ".red());
                }
            }
            if let Some(ref new) = diff.new_content {
                for line in new.lines() {
                    println!("{}{line}", "+ ".green());
                }
            }
            println!();
        }
    }

    Ok(())
}

async fn approve_review(client: &CowikiClient, id: &str, json: bool) -> Result<(), CliError> {
    client.approve_review(id).await?;

    if json {
        println!("{{\"ok\": true}}");
    } else {
        output::print_success(&format!("approved submission \"{id}\""));
    }

    Ok(())
}

async fn reject_review(client: &CowikiClient, id: &str, json: bool) -> Result<(), CliError> {
    client.reject_review(id).await?;

    if json {
        println!("{{\"ok\": true}}");
    } else {
        output::print_success(&format!("rejected submission \"{id}\""));
    }

    Ok(())
}

File: cli/src/commands/review.ts
import { Command } from 'commander';
import chalk from 'chalk';
import { loadConfig } from '../config.js';
import { CowikiClient } from '../client.js';
import { printSuccess, printError, printInfo, printJson, printTable } from '../output.js';
import { requireWorkspace } from '../shared.js';
import type { GlobalOpts } from '../shared.js';

export function registerReviewCommand(program: Command): void {
  const reviewCmd = program
    .command('review')
    .description('Review submissions');

  // ── review list ────────────────────────────────

  reviewCmd
    .command('list')
    .description('List review submissions')
    .option('--status <status>', 'Filter by status (pending/approved/rejected)')
    .action(async (opts, cmd) => {
      const globalOpts = (cmd.parent?.parent as Command)?.optsWithGlobals() as GlobalOpts;
      const config = loadConfig(globalOpts?.server)
      const ws = requireWorkspace(globalOpts.workspace);
      const client = new CowikiClient(config.baseUrl, config.apiKey);

      try {
        let submissions = await client.listReviews(ws);

        // Client-side status filtering
        if (opts.status) {
          const statusFilter = opts.status.toLowerCase();
          submissions = submissions.filter((s) => s.status.toLowerCase() === statusFilter);
        }

        if (globalOpts.json) {
          printJson(submissions);
        } else if (submissions.length === 0) {
          printInfo('no reviews found');
        } else {
          printTable(
            ['ID', 'USER', 'STATUS', 'SUMMARY', 'CREATED'],
            submissions.map((s) => {
              let statusColored: string;
              switch (s.status) {
                case 'pending': statusColored = chalk.yellow(s.status); break;
                case 'approved': statusColored = chalk.green(s.status); break;
                case 'rejected': statusColored = chalk.red(s.status); break;
                default: statusColored = s.status;
              }
              return [s.id, s.user_id, statusColored, s.summary, s.created_at];
            }),
          );
        }
      } catch (e) {
        printError(e instanceof Error ? e.message : String(e));
        process.exit(1);
      }
    });

  // ── review show ────────────────────────────────

  reviewCmd
    .command('show')
    .description('Show review details with diffs')
    .argument('<id>', 'Submission ID')
    .action(async (id, opts, cmd) => {
      const globalOpts = (cmd.parent?.parent as Command)?.optsWithGlobals() as GlobalOpts;
      const config = loadConfig(globalOpts?.server)
      const ws = requireWorkspace(globalOpts.workspace);
      const client = new CowikiClient(config.baseUrl, config.apiKey);

      try {
        const detail = await client.getReview(ws, id);

        if (globalOpts.json) {
          printJson(detail);
          return;
        }

        const sub = detail.submission;
        let statusColored: string;
        switch (sub.status) {
          case 'pending': statusColored = chalk.yellow(sub.status); break;
          case 'approved': statusColored = chalk.green(sub.status); break;
          case 'rejected': statusColored = chalk.red(sub.status); break;
          default: statusColored = sub.status;
        }

        console.log(`${chalk.bold('Submission:')} ${sub.id}`);
        console.log(`  User:     ${sub.user_id}`);
        console.log(`  Status:   ${statusColored}`);
        console.log(`  Summary:  ${sub.summary}`);
        console.log(`  Branch:   ${sub.source_branch}`);
        console.log(`  Created:  ${sub.created_at}`);
        if (sub.reviewed_by) console.log(`  Reviewed by: ${sub.reviewed_by}`);
        if (sub.reviewed_at) console.log(`  Reviewed at: ${sub.reviewed_at}`);
        console.log();

        if (!detail.diffs || detail.diffs.length === 0) {
          printInfo('no diffs');
        } else {
          for (const diff of detail.diffs) {
            console.log(`${chalk.yellow('---')} ${diff.path}`);
            if (diff.old_content) {
              for (const line of diff.old_content.split('\n')) {
                console.log(chalk.red(`- ${line}`));
              }
            }
            if (diff.new_content) {
              for (const line of diff.new_content.split('\n')) {
                console.log(chalk.green(`+ ${line}`));
              }
            }
            console.log();
          }
        }
      } catch (e) {
        const msg = e instanceof Error ? e.message : String(e);
        if (msg.includes('404')) {
          printError(`review not found: "${id}"`);
        } else {
          printError(msg);
        }
        process.exit(1);
      }
    });

  // ── review approve ─────────────────────────────

  reviewCmd
    .command('approve')
    .description('Approve a submission')
    .argument('<id>', 'Submission ID')
    .action(async (id, opts, cmd) => {
      const globalOpts = (cmd.parent?.parent as Command)?.optsWithGlobals() as GlobalOpts;
      const config = loadConfig(globalOpts?.server)
      const ws = requireWorkspace(globalOpts.workspace);
      const client = new CowikiClient(config.baseUrl, config.apiKey);

      try {
        await client.approveReview(ws, id);
        if (globalOpts.json) {
          console.log('{"ok":true}');
        } else {
          printSuccess(`approved submission "${id}"`);
        }
      } catch (e) {
        printError(e instanceof Error ? e.message : String(e));
        process.exit(1);
      }
    });

  // ── review reject ──────────────────────────────

  reviewCmd
    .command('reject')
    .description('Reject a submission')
    .argument('<id>', 'Submission ID')
    .action(async (id, opts, cmd) => {
      const globalOpts = (cmd.parent?.parent as Command)?.optsWithGlobals() as GlobalOpts;
      const config = loadConfig(globalOpts?.server)
      const ws = requireWorkspace(globalOpts.workspace);
      const client = new CowikiClient(config.baseUrl, config.apiKey);

      try {
        await client.rejectReview(ws, id);
        if (globalOpts.json) {
          console.log('{"ok":true}');
        } else {
          printSuccess(`rejected submission "${id}"`);
        }
      } catch (e) {
        printError(e instanceof Error ? e.message : String(e));
        process.exit(1);
      }
    });
}
